File: keyring/js/golden.test.mjs
// The JS twin's half of the keyring compatibility contract: replay
// ../testdata/golden.json — the same pinned vectors the Go test
// replays — through WebCrypto, plus round trips for the randomised
// directions. keyring.mjs imports ./crypto.mjs, which is not checked
// in beside it: the Go shim (js_test.go) materialises the serve-time
// sibling layout in a temp dir and runs this file there. To run by
// hand, copy crypto/js/crypto.mjs into keyring/js/ first, then
// `node --test js/golden.test.mjs` from keyring/ (Node ≥ 20).
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { ring, newSeed } from "./keyring.mjs";
import { unb64, b64, generate } from "./crypto.mjs";

const golden = JSON.parse(
  await readFile(fileURLToPath(new URL("../testdata/golden.json", import.meta.url)), "utf8"));

const r = ring(golden.namespace);

test("kass strings verbatim", () => {
  assert.equal(golden.strings.prf_salt, "kass/prf/v1");
  assert.equal(golden.strings.content_context, "kass/content/v1");
  assert.equal(golden.strings.wrap_context, "kass/wrap/v1");
  assert.equal(golden.strings.grant_context, "kass/grant/v1");
  assert.equal(r.prfSalt(), golden.strings.prf_salt);
});

test("contentKey reproduces the pinned key", async () => {
  const got = await r.contentKey(unb64(golden.seed));
  assert.equal(b64(got), golden.content_key);
});

test("wrapKey reproduces the pinned key", async () => {
  const got = await r.wrapKey(unb64(golden.prf));
  assert.equal(b64(got), golden.wrap_key);
});

test("unwrapSeed replays the pinned wrapped blob", async () => {
  const got = await r.unwrapSeed(unb64(golden.prf), unb64(golden.wrapped_seed));
  assert.equal(b64(got), golden.seed);
});

test("unwrapSeed refuses a wrong PRF output", async () => {
  const wrong = crypto.getRandomValues(new Uint8Array(32));
  await assert.rejects(() => r.unwrapSeed(wrong, unb64(golden.wrapped_seed)));
});

test("openGrant replays the pinned grant from the pkcs8 import", async () => {
  const got = await r.openGrant(unb64(golden.member.box_priv_pkcs8), unb64(golden.grant));
  assert.equal(b64(got), golden.content_key);
});

test("wrapSeed here unwraps here — device add is the same seed, new wrap", async () => {
  const seed = newSeed();
  const prfA = crypto.getRandomValues(new Uint8Array(32));
  const prfB = crypto.getRandomValues(new Uint8Array(32));
  const wrapA = await r.wrapSeed(prfA, seed);
  const wrapB = await r.wrapSeed(prfB, seed);
  assert.notDeepEqual(wrapA, wrapB);
  assert.deepEqual(await r.unwrapSeed(prfA, wrapA), seed);
  assert.deepEqual(await r.unwrapSeed(prfB, wrapB), seed);
});

test("grant here opens here with a JWK member key", async () => {
  const member = await generate();
  const key = crypto.getRandomValues(new Uint8Array(32));
  const sealed = await r.grant(member.boxPub, key);
  assert.deepEqual(await r.openGrant(member.boxPrivJwk, sealed), key);
});

test("a grant is namespaced", async () => {
  const member = await generate();
  const key = crypto.getRandomValues(new Uint8Array(32));
  const sealed = await r.grant(member.boxPub, key);
  const other = ring("app");
  await assert.rejects(() => other.openGrant(member.boxPrivJwk, sealed));
});

test("blobKey matches the Go derivation, pinned by value", async () => {
  // Not in golden.json (that file is hash-pinned, wire-format-forever);
  // pinned instead against the Go side's deterministic output for
  // seed = 0x07 x32, ring "kass", name "servers" — see
  // keyring/blobkey_test.go.
  const seed = new Uint8Array(32).fill(7);
  const got = await r.blobKey(seed, "servers");
  assert.equal(
    Buffer.from(got).toString("hex"),
    "f3b239c47fc8cec5c20588083f93c73640b6b75ce8a1473576d93ebe172d2d43");
});

test("blobKey isolates names", async () => {
  const seed = new Uint8Array(32).fill(7);
  const a = Buffer.from(await r.blobKey(seed, "servers")).toString("hex");
  const b = Buffer.from(await r.blobKey(seed, "drafts")).toString("hex");
  assert.notEqual(a, b);
});

File: keyring/golden_test.go
package keyring

import (
	"bytes"
	"crypto/ecdh"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"amadan.net/rastrillo/rastrillo/crypto"
)

// goldenFile is the shape of testdata/golden.json — this package's
// pinned cross-implementation fixture, shared with the generator
// (golden_gen_test.go) so writer and reader cannot drift. Wrap and
// Grant mint random IVs and ephemeral keys, so their outputs are not
// vectorable; the file pins the deterministic directions — purpose
// derivation, UnwrapSeed of a pinned blob, OpenGrant of a pinned
// grant, the ns="kass" strings verbatim — crypto's own golden pattern.
type goldenFile struct {
	Provenance string `json:"_provenance"`
	Namespace  string `json:"namespace"`
	Strings    struct {
		PRFSalt string `json:"prf_salt"`
		Content string `json:"content_context"`
		Wrap    string `json:"wrap_context"`
		Grant   string `json:"grant_context"`
	} `json:"strings"`
	Seed        string `json:"seed"`
	ContentKey  string `json:"content_key"`
	PRF         string `json:"prf"`
	WrapKey     string `json:"wrap_key"`
	WrappedSeed string `json:"wrapped_seed"`
	Member      struct {
		BoxPrivD     string `json:"box_priv_d"`
		BoxPrivPKCS8 string `json:"box_priv_pkcs8"`
		BoxPub       string `json:"box_pub"`
	} `json:"member"`
	Grant string `json:"grant"`
}

func loadGolden(t *testing.T) goldenFile {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("testdata", "golden.json"))
	if err != nil {
		t.Fatalf("read testdata/golden.json: %v", err)
	}
	var g goldenFile
	if err := json.Unmarshal(raw, &g); err != nil {
		t.Fatalf("unmarshal testdata/golden.json: %v", err)
	}
	return g
}

func mustB64(t *testing.T, s string) []byte {
	t.Helper()
	b, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		t.Fatalf("base64 decode %q: %v", s, err)
	}
	return b
}

// TestGoldenVectors replays the pinned fixture against this
// implementation — the compatibility contract that lets kass adopt
// Ring{"kass"} without a data migration.
func TestGoldenVectors(t *testing.T) {
	g := loadGolden(t)
	r := Ring{Namespace: g.Namespace}

	t.Run("kass strings verbatim", func(t *testing.T) {
		if g.Strings.PRFSalt != "kass/prf/v1" || g.Strings.Content != "kass/content/v1" ||
			g.Strings.Wrap != "kass/wrap/v1" || g.Strings.Grant != "kass/grant/v1" {
			t.Fatalf("pinned strings drifted: %+v", g.Strings)
		}
		if got := r.PRFSalt(); got != g.Strings.PRFSalt {
			t.Fatalf("PRFSalt() = %q, want pinned %q", got, g.Strings.PRFSalt)
		}
	})

	t.Run("purpose derivation reproduces the pinned keys", func(t *testing.T) {
		seed := mustB64(t, g.Seed)
		if !bytes.Equal(r.ContentKey(seed), mustB64(t, g.ContentKey)) {
			t.Fatal("ContentKey(seed) does not match pinned content_key")
		}
		if !bytes.Equal(crypto.Derive(seed, g.Strings.Content), mustB64(t, g.ContentKey)) {
			t.Fatal("pinned content_context does not derive the pinned content_key — the context string drifted")
		}
		prf := mustB64(t, g.PRF)
		if !bytes.Equal(r.WrapKey(prf), mustB64(t, g.WrapKey)) {
			t.Fatal("WrapKey(prf) does not match pinned wrap_key")
		}
		if !bytes.Equal(crypto.Derive(prf, g.Strings.Wrap), mustB64(t, g.WrapKey)) {
			t.Fatal("pinned wrap_context does not derive the pinned wrap_key — the context string drifted")
		}
	})

	t.Run("UnwrapSeed replays the pinned wrapped blob", func(t *testing.T) {
		got, err := r.UnwrapSeed(mustB64(t, g.PRF), mustB64(t, g.WrappedSeed))
		if err != nil {
			t.Fatalf("UnwrapSeed: %v", err)
		}
		if !bytes.Equal(got, mustB64(t, g.Seed)) {
			t.Fatal("UnwrapSeed(pinned) != pinned seed")
		}
	})

	t.Run("OpenGrant replays the pinned grant", func(t *testing.T) {
		priv, err := hex.DecodeString(g.Member.BoxPrivD)
		if err != nil {
			t.Fatalf("decode box_priv_d: %v", err)
		}
		got, err := r.OpenGrant(priv, mustB64(t, g.Grant))
		if err != nil {
			t.Fatalf("OpenGrant: %v", err)
		}
		if !bytes.Equal(got, mustB64(t, g.ContentKey)) {
			t.Fatal("OpenGrant(pinned) != pinned content_key")
		}
	})

	t.Run("the grant context string is pinned through crypto.Open", func(t *testing.T) {
		raw, err := hex.DecodeString(g.Member.BoxPrivD)
		if err != nil {
			t.Fatalf("decode box_priv_d: %v", err)
		}
		boxPriv, err := ecdh.P256().NewPrivateKey(raw)
		if err != nil {
			t.Fatalf("reconstruct box key: %v", err)
		}
		got, err := crypto.Open(&crypto.Keypair{BoxPriv: boxPriv}, g.Strings.Grant, mustB64(t, g.Grant))
		if err != nil {
			t.Fatalf("crypto.Open under pinned grant_context: %v", err)
		}
		if !bytes.Equal(got, mustB64(t, g.ContentKey)) {
			t.Fatal("crypto.Open under pinned grant_context != pinned content_key")
		}
	})

	t.Run("round trips for the randomised directions", func(t *testing.T) {
		seed, err := NewSeed()
		if err != nil {
			t.Fatalf("NewSeed: %v", err)
		}
		prf, err := crypto.NewKey()
		if err != nil {
			t.Fatalf("NewKey: %v", err)
		}
		wrapped, err := r.WrapSeed(prf, seed)
		if err != nil {
			t.Fatalf("WrapSeed: %v", err)
		}
		gotSeed, err := r.UnwrapSeed(prf, wrapped)
		if err != nil {
			t.Fatalf("UnwrapSeed: %v", err)
		}
		if !bytes.Equal(gotSeed, seed) {
			t.Fatal("WrapSeed/UnwrapSeed round trip lost the seed")
		}

		member, err := crypto.Generate()
		if err != nil {
			t.Fatalf("Generate: %v", err)
		}
		key, err := crypto.NewKey()
		if err != nil {
			t.Fatalf("NewKey: %v", err)
		}
		sealed, err := r.Grant(member.BoxPub(), key)
		if err != nil {
			t.Fatalf("Grant: %v", err)
		}
		gotKey, err := r.OpenGrant(member.BoxPriv.Bytes(), sealed)
		if err != nil {
			t.Fatalf("OpenGrant: %v", err)
		}
		if !bytes.Equal(gotKey, key) {
			t.Fatal("Grant/OpenGrant round trip lost the content key")
		}
	})
}

// TestVectorsFileUntouched pins the vectors file byte-for-byte:
// "fix implementations, not vectors" — grant bytes are wire-format-
// forever, so no change to this package may rewrite the spec it is
// tested against (eventlog's TestMergeVectorsFileUntouched pattern).
func TestVectorsFileUntouched(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join("testdata", "golden.json"))
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	const want = "008df2f8291ad0e248dac69d48f4f6544905539cad910bdbfc549ff167180b62"
	if got := fmt.Sprintf("%x", sha256.Sum256(raw)); got != want {
		t.Fatalf("golden.json changed (sha256 %s, want %s); fix implementations, not vectors", got, want)
	}
}
